### DISCUSSION TITLE
Problem with GDAL package and readthedocs

### DISCUSSION THREAD

**Discussion by PratyushTripathy** [pypa/setuptools#3083] [Q&A]
I am trying to set up readthedocs for the [pyrsgis](https://github.com/PratyushTripathy/pyrsgis) Python package. The documentation used to [build fine](https://pyrsgis.readthedocs.io/en/master/) until recently when it started failing. Documentation can only be built with GDAL 2.2.3 since readthedocs supports this version only. However, Setuptools does not seem to work with that. Please see [this suggestion](https://github.com/readthedocs/readthedocs.org/issues/8365#issuecomment-994823157) and [this](https://github.com/readthedocs/readthedocs.org/issues/8365#issuecomment-995212357) by readthedocs development team. They suggest that it can be solved on setuptools' end. Please advise. Thanks!

**Comment by abravalheri** [comment]
Hi @PratyushTripathy,

I had a look on the linked issue and I noticed that you are pinning setuptools 58.0.0.

According to the [changelog](https://setuptools.pypa.io/en/latest/history.html#v58-0-0), this version already had dropped support for `2to3`. Have you tried to pin an older version?

Do you know if the latest setuptools works with the latest GDAL localy (I mean, not on RTD)?

**Reply by PratyushTripathy** [reply]
Hi @abravalheri,

Thanks for pointing this out. I pinned v57.5.0 now, which is just before the support for `2to3` was dropped. But that did not work. The readthedocs [build failed](https://readthedocs.org/projects/pyrsgis/builds/16006704/). Copying @astrojuanlu

> Do you know if the latest setuptools works with the latest GDAL localy (I mean, not on RTD)?

Have not tried this.

**Reply by astrojuanlu** [reply]
Hi @PratyushTripathy, please try building GDAL 2.2.3 with setuptools 57.5.0 locally, so we make sure it works.

A couple of things about RTD (apologies for dumping this information on the setuptools discussions):

> Documentation can only be built with GDAL 2.2.3 since readthedocs supports this version only.

A few months after we discussed [that](https://github.com/readthedocs/readthedocs.org/issues/8365#issuecomment-886399073), RTD [introduced base images based on Ubuntu 20.04](https://blog.readthedocs.com/new-build-specification/), which [ship libgdal-dev 3.0.4](https://packages.ubuntu.com/focal/libgdal-dev). Therefore, maybe you can work around these issues by upgrading to a newer base image and GDAL version. If you attempt that, it will be better if you open an issue on the RTD tracker (sorry for bouncing you back and forth).

>  I pinned v57.5.0 now, which is just before the support for 2to3 was dropped. But that did not work.

This has to do with how RTD installs dependencies. Yes, [you added a `docs_requirements_setuptools.txt` with that version pinned](https://github.com/PratyushTripathy/pyrsgis/commit/c5607cf802cfef4cdebcfe67d4870f643b5e5afe) ✔️ but [your `.readthedocs.yaml` is not using it](https://github.com/PratyushTripathy/pyrsgis/blob/961d6f4cc0ccb21a8d66d121d15fedc8e3f05a73/.readthedocs.yaml#L24-L30), and therefore the installation uses the setuptools version pinned by RTD, which is `setuptools<58.3.0` (see discussion at https://github.com/pypa/setuptools/discussions/2904). You'd have to do something like

```
python:
  install:
      - requirements: docs_requirements_setuptools.txt
      - requirements: docs_requirements.txt
      - method: pip
        path: .
        extra_requirements:
           - tests
           - docs
```

again, if you have trouble with this, better open a separate issue on the RTD tracker.
